### ISSUE TITLE
Please don't define _GNU_SOURCE in util.h

### ISSUE THREAD

**Issue by rbalint** [libevent/libevent#460]
Forwarded from:
https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=632490

As discussed in the original bug it would be nicer if libevent would not require applications using libevent to have _GNU_SOURCE defined.

**Comment by azat** [comment]
Refs: #773

**Comment by ploxiln** [comment]
ideally, `_GNU_SOURCE` would only be defined at the top of c files in which it is needed ... it looks like it is intended for use with `netdb.h` which provides `getaddrinfo()` `gethostbyname()` etc, but right now linux man pages say:

```
   Feature Test Macro Requirements for glibc (see feature_test_macros(7)):

       getaddrinfo(), freeaddrinfo(), gai_strerror():
           _POSIX_C_SOURCE >= 1 || _XOPEN_SOURCE || _POSIX_SOURCE
```

```
       gethostbyname2(), gethostent_r(), gethostbyaddr_r(), gethostbyname_r(), gethostbyname2_r():
           _BSD_SOURCE || _SVID_SOURCE

       herror(), hstrerror():
           Since glibc 2.8:
               _BSD_SOURCE || _SVID_SOURCE
           Before glibc 2.8:
               none

       h_errno:
           Since glibc 2.12:
               _BSD_SOURCE || _SVID_SOURCE ||
                   (_POSIX_C_SOURCE < 200809L && _XOPEN_SOURCE < 700)
           Before glibc 2.12:
               none
```

(there is a reference to `h_errno` in `evutil.c`)

It could be annoying to figure out exactly which old and diverse systems need `_GNU_SOURCE` and for what ...

**Comment by ngie-eign** [comment]
> It could be annoying to figure out exactly which old and diverse systems need `_GNU_SOURCE` and for what ...

Supporting a decade old POSIX spec conforming Linux (2008) or older doesn't really make sense :/.

**Comment by azat** [comment]
> `_GNU_SOURCE` would only be defined at the top of c files

well basically this is a good idea, but we need deep testing for this

since we define `_GNU_SOURCE` not only in `<event2/util.h>` but also in `<event2/event-config.h>` and `<evconfig-private.h>`, while later does not included from public headers, the former does, and just removing that define out will not helps here

**Comment by azat** [comment]
And the interesting bit here is cmake, since it does not define `_GNU_SOURCE` in `event2/event-config.h` and everything works when cmake is used for building, so I guess we can safely remove them.

**Comment by azat** [comment]
Nope, [it](https://formulae.brew.sh/api/formula/libevent.json) uses autotools (checked [this](https://homebrew.bintray.com/bottles/libevent-2.1.8.mojave.bottle.tar.gz) one.

And installed headers looks good, no `util.h` in `/include/`

Although, this is another story, but since this contained in commit message, I guess it is better to figure out where this error came from to update message, what do you think?

**Comment by ploxiln** [comment]
```
evthread_pthread.c
29:/* With glibc we need to define _GNU_SOURCE to get PTHREAD_MUTEX_RECURSIVE.
```

**Comment by ploxiln** [comment]
buffer.c
```
#ifdef EVENT__HAVE_VASPRINTF
/* If we have vasprintf, we need to define _GNU_SOURCE before we include
 * stdio.h.  This comes from evconfig-private.h.
 */
#endif
```

not clear why, or if libevent uses vasprintf() anywhere, indirectly some how?

**Comment by azat** [comment]
> ``` With glibc we need to define _GNU_SOURCE to get PTHREAD_MUTEX_RECURSIVE.

```
```shell
gcc -c -o /dev/null -xc - <<<$'#include <pthread.h>\nint foo() { return PTHREAD_MUTEX_RECURSIVE; }'
```

Outdated comment? I guess

> EVENT__HAVE_VASPRINTF

Cannot find anything, will clean this up later.

Thanks for catching this, but these are not headers anyway.

**Comment by azat** [comment]
[Here](https://gist.github.com/azat/77a539b1c21ea54b3848500d48f0d435) is the list of used extensions

**Comment by ploxiln** [comment]
yeah, I was just trying to figure out what `_GNU_SOURCE` was originally intended for, to see if it is still needed for anything, and how small the scope it is needed for is ...

(I found a couple more uses that were added then removed, first was `F_SETSIG`, then `fnmatch()`, and then there's this non-specific commit: 4d48cf61a2734aa37a7a5904f5c1454a62aa943a)

**Comment by azat** [comment]
Well basically what we want to do is define `-D_GNU_SOURCE` in `autotools` during compilation (like we do in `cmake`) and do not write it to the `event-config.h`.

And remove every definition of `_GNU_SOURCE` in c modules.

How that sounds?

**Comment by ploxiln** [comment]
Yes, that should work fine.

In my own projects I prefer to use `-D_XOPEN_SOURCE=600` and then just `#define _GNU_SOURCE` at the top of a c file (before all includes) if it really needs it for some function, because there are a rare few functions like `strerror_r()` where you don't get the cross-platform "XSI-compliant" version if `_GNU_SOURCE` is defined. Thus the attempt to have `_GNU_SOURCE` defined only for a couple small c files, or not at all.

But for libevent is compatible with more older platforms than I've ever needed to ensure compatibility with for my stuff, and has worked with `_GNU_SOURCE` for a long while, so just `-D_GNU_SOURCE` does seem to be the way to go here.

**Comment by azat** [comment]
Looks like this had been fixed in 5f87be42f0ae0126938624a1419a572607078217
